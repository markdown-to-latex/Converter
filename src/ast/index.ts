export * as node from './node';
export * as parsing from './parsing';
export * from './nodes';
