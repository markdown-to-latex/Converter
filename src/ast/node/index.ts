export * from './extension';
export * from './function';
export * from './position';
export * from './struct';
