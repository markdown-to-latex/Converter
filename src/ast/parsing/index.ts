import {
    createStartEndTextPos,
    FileNode,
    NodeE,
    NodeType,
    RawNode,
    RawNodeType,
    StartEndTextPosition,
} from '../node';
import path from 'path';
import { StringE } from '../../extension/string';
import { tokenize, tokensToNode } from './tokenizer';

export function fullContentPos(
    content: string | StringE,
): StartEndTextPosition {
    const contentE = StringE.from(content);

    const lines = contentE.splitE(/\r?\n/);
    return createStartEndTextPos(
        1,
        1,
        lines.length,
        lines[lines.length - 1].length + 1,
    );
}

function parseFile(content: string, filePath: string) {
    const contentNode: RawNode = {
        type: RawNodeType.Raw,
        parent: null,
        text: content,
        pos: {
            start: 0,
            end: content.length,
        },
    };

    const fileNode: FileNode = {
        type: NodeType.File,
        parent: null,
        children: [contentNode],
        pos: {
            start: 0,
            end: content.length,
        },
        raw: content,
        path: path.resolve(filePath),
    };

    contentNode.parent = fileNode;

    let nodeE = NodeE.from(fileNode);
}
