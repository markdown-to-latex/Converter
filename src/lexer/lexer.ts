import { Token, TokenListContainer } from './tokens';
import * as marked from 'marked';
import {
    processTokenChildren,
    processTokenIfVisitorExists,
    TokenWithChildren,
} from './visitors';

function parseTokens(tokens: Readonly<Token[]>): Token[] {
    const resultList: Token[] = [];
    for (const token of tokens) {
        resultList.push(...processTokenIfVisitorExists(token));
    }
    for (const token of resultList) {
        (['tokens', 'items'] as (keyof TokenWithChildren)[]).forEach(key =>
            processTokenChildren(key, token as TokenWithChildren),
        );
    }

    return resultList;
}

export function lexer(
    text: string,
    options?: marked.MarkedOptions,
): TokenListContainer {
    const originalTokenList: Readonly<TokenListContainer> = marked.lexer(
        text,
        options,
    );
    const links = originalTokenList.links;

    const resultList: TokenListContainer = parseTokens(
        originalTokenList,
    ) as TokenListContainer;
    resultList.links = links;

    return resultList;
}
