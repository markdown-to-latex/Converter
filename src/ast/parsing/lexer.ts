import * as node from '../node';
import {
    CodeNode,
    copyTextPosition,
    createStartEndPos,
    Node,
    NodeType,
    RawNode,
    RawNodeE,
    RawNodeType,
    TextPosition,
    textPositionG,
    textPositionGEq,
} from '../node';
import {
    DiagnoseErrorType,
    DiagnoseList,
    diagnoseListHasSeverity,
    DiagnoseSeverity,
} from '../../diagnose';

export const enum LexemeType {
    // Space,
    // Paragraph,
    // Heading,
    // Text,
    Code,
}

const parserPriorities: LexemeType[] = [
    LexemeType.Code,
    // LexemeType.Heading,
    // LexemeType.Paragraph,
    // LexemeType.Text,
    // LexemeType.Space,
];

interface LexemeTypeToNodeType {
    [LexemeType.Code]: node.CodeNode;
    // [LexemeType.Heading]: node.HeadingNode;
    // [LexemeType.Paragraph]: node.ParagraphNode;
    // [LexemeType.Text]: node.TextNode;
    // [LexemeType.Space]: node.SpaceNode;
}

class FatalError extends Error {
}

export function applyVisitors(nodes: Readonly<Node[]>): [Node[], DiagnoseList] {
    let allDiags: DiagnoseList = [];

    let newNodes = [...nodes];
    try {
        for (const type of parserPriorities) {
            newNodes = newNodes.flatMap(node => {
                if (node.type !== RawNodeType.Raw) {
                    return [node];
                }

                let [nodes, diags] = applyParser(node as RawNode, type);
                allDiags.push(...diags);

                if (diagnoseListHasSeverity(diags, DiagnoseSeverity.Fatal)) {
                    throw new FatalError();
                }

                return nodes;
            });
        }
    } catch (e) {
        if (e instanceof FatalError) {
            return [[], allDiags];
        } else {
            throw e;
        }
    }

    return [newNodes, allDiags];
}

function applyParser(
    node: Readonly<RawNode>,
    type: LexemeType,
): [Node[], DiagnoseList] {
    const nodeE = new RawNodeE(node);

    const diags: DiagnoseList = [];
    const newNodes: Node[] = [];
    let position: TextPosition = copyTextPosition(node.pos.start);

    const parsed = parsers[type](nodeE);
    for (const node of parsed) {
        if (!textPositionGEq(position, node.pos.start)) {
            newNodes.push({
                type: RawNodeType.Raw,
                pos: {
                    start: copyTextPosition(position),
                    end: copyTextPosition(node.pos.start),
                },
                text: nodeE.text.slicePosition(
                    nodeE.n.pos.start,
                    position,
                    node.pos.start,
                ).s,
                parent: nodeE.n.parent,
            } as RawNode);
        }

        newNodes.push(node);
        position = copyTextPosition(node.pos.end);
    }

    if (!textPositionGEq(position, node.pos.end)) {
        newNodes.push({
            type: RawNodeType.Raw,
            pos: {
                start: copyTextPosition(position),
                end: copyTextPosition(node.pos.end),
            },
            text: nodeE.text.slicePosition(
                nodeE.n.pos.start,
                position,
                node.pos.end,
            ).s,
            parent: nodeE.n.parent,
        } as RawNode);
    } else if (textPositionG(position, node.pos.end)) {
        diags.push({
            errorType: DiagnoseErrorType.ApplyParserError,
            severity: DiagnoseSeverity.Fatal,
            pos: {
                start: node.pos.end,
                end: position,
            },
            message: 'Parser position overflow',
            filePath: nodeE.parentFile?.path ?? 'null',
        });
    }

    return [newNodes, diags];
}

type Visitor<T> = (node: RawNodeE) => T[];
const parsers: {
    [key in LexemeType]: Visitor<LexemeTypeToNodeType[key]>;
} = {
    [LexemeType.Code]: node => {
        const lines = node.text.lines;

        const codeLexemes = lines
            .map((v, i) => ({
                offLine: i,
                absLine: node.n.pos.start.line + i,
                match: v.s.match(/\s*`{3,}/),
            }))
            .filter(v => v.match?.length);
        if (codeLexemes.length % 2 !== 0) {
            const lastLexeme = codeLexemes[codeLexemes.length - 1];
            const absLine = lastLexeme.absLine;
            throw new Error(
                `Unable to find closing quotes for block code. ` +
                `Began at ${absLine} at file TODO`,
            );
        }

        const codeNodes: CodeNode[] = [];
        for (let i = 0; i < codeLexemes.length; i += 2) {
            const startLexeme = codeLexemes[i];
            const endLexeme = codeLexemes[i + 1];

            const text = lines
                .slice(startLexeme.offLine + 1, endLexeme.offLine)
                .join('\n');
            codeNodes.push({
                type: NodeType.Code,
                pos: createStartEndPos(
                    startLexeme.absLine,
                    1,
                    endLexeme.absLine,
                    lines[endLexeme.offLine].length,
                ),
                text: text,
                parent: node.n.parent,
                lang: null,
                codeBlockStyle: null,
            });
        }

        return codeNodes;
    },
};
