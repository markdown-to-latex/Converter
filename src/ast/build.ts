import { FileNode, getNodeAllChildren, Node, NodeType } from './nodes';
import { TokenListContainer } from '../lexer/tokens';
import { astProcessTokenList } from './visitors';

function updateASTParents(ast: Node) {
    interface NodeInfo {
        parent: Node | null;
        node: Node;
    }

    const nodeQueue: NodeInfo[] = [];
    nodeQueue.push({
        parent: null,
        node: ast,
    });

    // In-depth AST traverse
    while (nodeQueue.length !== 0) {
        const info = nodeQueue.pop()!;

        // Update parent
        info.node.parent = info.parent;

        nodeQueue.push(
            ...getNodeAllChildren(
                info.node as Parameters<typeof getNodeAllChildren>[0],
            ).map(node => ({
                parent: info.node,
                node: node,
            })),
        );
    }
}

export interface MarkdownASTData {
    filepath?: string;
}

export function buildMarkdownAST(
    tokens: Readonly<TokenListContainer>,
    options: Readonly<MarkdownASTData>,
): FileNode {
    const ast: FileNode = {
        type: NodeType.File,
        parent: null,
        children: astProcessTokenList(tokens),
        path: options.filepath ?? '__not-specified__',
    };
    updateASTParents(ast);

    return ast;
}
