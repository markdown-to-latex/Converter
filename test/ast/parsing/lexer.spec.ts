import {
    CodeNode, CodeSpanNode,
    FileNode, LinkNode,
    NodeType, OpCodeNode,
    RawNode,
    RawNodeType, TableNode,
} from '../../../src/ast/node';
import {fullContentPos} from '../../../src/ast/parsing';
import {applyVisitors} from '../../../src/ast/parsing/lexer';

function rawNodeTemplate(content: string): RawNode {
    const fileNode: FileNode = {
        type: NodeType.File,
        parent: null,
        pos: {
            start: 0,
            end: content.length,
        },
        raw: content,
        path: 'test.md',
        children: [],
    };

    const rawNode: RawNode = {
        type: RawNodeType.Raw,
        parent: fileNode,
        pos: {
            start: 0,
            end: content.length,
        },
        text: content,
    };
    fileNode.children.push(rawNode);

    return rawNode;
}

describe('code block lexer check', () => {
    test('Complex raw node', () => {
        const rawNode = rawNodeTemplate(`Sample text
\`\`\`
Code block
\`\`\`
New sample text
`);
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);

        let node = nodes[1] as CodeNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.Code);
        expect(node.text).toEqual('Code block');
        expect(rawNode.text.slice(node.pos.start, node.pos.end)).toEqual(
            '```\nCode block\n```',
        );
        expect(node.parent).toEqual(rawNode.parent);

        expect(nodes).toMatchSnapshot();
    });
    test('With language', () => {
        const rawNode = rawNodeTemplate(`Sample text
\`\`\`test-language  
Code block
\`\`\`
New sample text
`);
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[1] as CodeNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.Code);
        expect(node.lang).toEqual('test-language');
    });

    test('Error raw node', () => {
        const rawNode = rawNodeTemplate(`Sample text
\`\`\`
Code block
New sample text
`);
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(1);
        expect(diagnostic).toMatchSnapshot();
    });
});

describe('link check', () => {
    test('Simple Link in text', () => {
        const rawNode = rawNodeTemplate('Hello [ti t le](li-nk) text');
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[1] as LinkNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.Link);

        expect(nodes).toMatchSnapshot();
    })

    test('Link with inner md in text', () => {
        const rawNode = rawNodeTemplate('Hello [ti \n`co][de` le](li-nk) text');
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[1] as LinkNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.Link);
        let codeSpan = node.children[1] as CodeSpanNode;
        expect(codeSpan).not.toBeUndefined();
        expect(codeSpan.type).toEqual(NodeType.CodeSpan);

        expect(nodes).toMatchSnapshot();
    })
})

describe('macros parsing', () => {
    test('Complex macros with name, label, pos and key args', () => {
        const rawNode = rawNodeTemplate('!Macro[label-text](pos arg 1)(`pos arg` 2)(@keyArgName argName)');
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[0] as OpCodeNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.OpCode);

        expect(nodes).toMatchSnapshot();
    })

    test('Key is defined wrongly', () => {
        const rawNode = rawNodeTemplate('!Macro(@key)(@)');
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[0] as OpCodeNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.OpCode);

        expect(nodes).toMatchSnapshot();
    })

    test('Multiple keys error', () => {
        const rawNode = rawNodeTemplate('!Macro(@key value)(@key another value)');
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(1);

        const errText = rawNode.text.slice(diagnostic[0].pos.start.absolute, diagnostic[0].pos.end.absolute);
        expect(errText).toEqual('key')
    })

    test('Brackets mismatch', () => {
        const rawNode = rawNodeTemplate('!Macro[(@key ]value)(@key another value)');
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[0] as OpCodeNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.OpCode);

        expect(nodes).toMatchSnapshot();
    })
})

describe('table parsing', () => {
    test('Simple table', () => {
        const rawNode = rawNodeTemplate(`| Column 1 | Column 2 |
| -------- | -------- |
| Value **1** | Value \`2\` |
|Value [with](link) | |`)
        const {nodes, diagnostic} = applyVisitors([rawNode]);
        expect(diagnostic).toHaveLength(0);
        let node = nodes[0] as TableNode;
        expect(node).not.toBeUndefined();
        expect(node.type).toEqual(NodeType.Table);

        expect(nodes).toMatchSnapshot();
    })
})