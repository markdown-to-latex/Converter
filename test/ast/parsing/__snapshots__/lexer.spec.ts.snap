// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code block lexer check Complex raw node 1`] = `
Array [
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 47,
            "start": 0,
          },
          "text": "Sample text
\`\`\`
Code block
\`\`\`
New sample text
",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 47,
        "start": 0,
      },
      "raw": "Sample text
\`\`\`
Code block
\`\`\`
New sample text
",
      "type": "File",
    },
    "pos": Object {
      "end": 11,
      "start": 0,
    },
    "text": "Sample text",
    "type": "Text",
  },
  Object {
    "codeBlockStyle": null,
    "lang": null,
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 47,
            "start": 0,
          },
          "text": "Sample text
\`\`\`
Code block
\`\`\`
New sample text
",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 47,
        "start": 0,
      },
      "raw": "Sample text
\`\`\`
Code block
\`\`\`
New sample text
",
      "type": "File",
    },
    "pos": Object {
      "end": 30,
      "start": 12,
    },
    "text": "Code block",
    "type": "Code",
  },
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 47,
            "start": 0,
          },
          "text": "Sample text
\`\`\`
Code block
\`\`\`
New sample text
",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 47,
        "start": 0,
      },
      "raw": "Sample text
\`\`\`
Code block
\`\`\`
New sample text
",
      "type": "File",
    },
    "pos": Object {
      "end": 46,
      "start": 31,
    },
    "text": "New sample text",
    "type": "Text",
  },
]
`;

exports[`code block lexer check Error raw node 1`] = `
Array [
  Object {
    "errorType": 0,
    "filePath": "test.md",
    "message": "Unable to find closing quotes for block code",
    "pos": Object {
      "end": Object {
        "absolute": 15,
        "column": 1,
        "line": 5,
      },
      "start": Object {
        "absolute": 12,
        "column": 1,
        "line": 2,
      },
    },
    "severity": "ERROR",
  },
]
`;

exports[`link check Link with inner md in text 1`] = `
Array [
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 35,
            "start": 0,
          },
          "text": "Hello [ti 
\`co][de\` le](li-nk) text",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 35,
        "start": 0,
      },
      "raw": "Hello [ti 
\`co][de\` le](li-nk) text",
      "type": "File",
    },
    "pos": Object {
      "end": 6,
      "start": 0,
    },
    "text": "Hello ",
    "type": "Text",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "pos": Object {
          "end": 10,
          "start": 7,
        },
        "text": "ti ",
        "type": "Text",
      },
      Object {
        "parent": [Circular],
        "pos": Object {
          "end": 19,
          "start": 11,
        },
        "text": "co][de",
        "type": "CodeSpan",
      },
      Object {
        "children": Array [],
        "parent": [Circular],
        "pos": Object {
          "end": 22,
          "start": 19,
        },
        "text": " le",
        "type": "Text",
      },
    ],
    "href": "li-nk",
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 35,
            "start": 0,
          },
          "text": "Hello [ti 
\`co][de\` le](li-nk) text",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 35,
        "start": 0,
      },
      "raw": "Hello [ti 
\`co][de\` le](li-nk) text",
      "type": "File",
    },
    "pos": Object {
      "end": 30,
      "start": 6,
    },
    "title": "",
    "type": "Link",
  },
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 35,
            "start": 0,
          },
          "text": "Hello [ti 
\`co][de\` le](li-nk) text",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 35,
        "start": 0,
      },
      "raw": "Hello [ti 
\`co][de\` le](li-nk) text",
      "type": "File",
    },
    "pos": Object {
      "end": 35,
      "start": 30,
    },
    "text": " text",
    "type": "Text",
  },
]
`;

exports[`link check Simple Link in text 1`] = `
Array [
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 27,
            "start": 0,
          },
          "text": "Hello [ti t le](li-nk) text",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 27,
        "start": 0,
      },
      "raw": "Hello [ti t le](li-nk) text",
      "type": "File",
    },
    "pos": Object {
      "end": 6,
      "start": 0,
    },
    "text": "Hello ",
    "type": "Text",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "pos": Object {
          "end": 14,
          "start": 7,
        },
        "text": "ti t le",
        "type": "Text",
      },
    ],
    "href": "li-nk",
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 27,
            "start": 0,
          },
          "text": "Hello [ti t le](li-nk) text",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 27,
        "start": 0,
      },
      "raw": "Hello [ti t le](li-nk) text",
      "type": "File",
    },
    "pos": Object {
      "end": 22,
      "start": 6,
    },
    "title": "",
    "type": "Link",
  },
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 27,
            "start": 0,
          },
          "text": "Hello [ti t le](li-nk) text",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 27,
        "start": 0,
      },
      "raw": "Hello [ti t le](li-nk) text",
      "type": "File",
    },
    "pos": Object {
      "end": 27,
      "start": 22,
    },
    "text": " text",
    "type": "Text",
  },
]
`;

exports[`macros parsing Brackets mismatch 1`] = `
Array [
  Object {
    "keyArgs": Object {},
    "label": "(@key ",
    "opcode": "Macro",
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 40,
            "start": 0,
          },
          "text": "!Macro[(@key ]value)(@key another value)",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 40,
        "start": 0,
      },
      "raw": "!Macro[(@key ]value)(@key another value)",
      "type": "File",
    },
    "pos": Object {
      "end": 14,
      "start": 0,
    },
    "posArgs": Array [],
    "type": "OpCode",
  },
  Object {
    "children": Array [],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 40,
            "start": 0,
          },
          "text": "!Macro[(@key ]value)(@key another value)",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 40,
        "start": 0,
      },
      "raw": "!Macro[(@key ]value)(@key another value)",
      "type": "File",
    },
    "pos": Object {
      "end": 40,
      "start": 14,
    },
    "text": "value)(@key another value)",
    "type": "Text",
  },
]
`;

exports[`macros parsing Complex macros with name, label, pos and key args 1`] = `
Array [
  Object {
    "keyArgs": Object {
      "keyArgName": Array [
        Object {
          "children": Array [],
          "parent": [Circular],
          "pos": Object {
            "end": 62,
            "start": 55,
          },
          "text": "argName",
          "type": "Text",
        },
      ],
    },
    "label": "label-text",
    "opcode": "Macro",
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 63,
            "start": 0,
          },
          "text": "!Macro[label-text](pos arg 1)(\`pos arg\` 2)(@keyArgName argName)",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 63,
        "start": 0,
      },
      "raw": "!Macro[label-text](pos arg 1)(\`pos arg\` 2)(@keyArgName argName)",
      "type": "File",
    },
    "pos": Object {
      "end": 63,
      "start": 0,
    },
    "posArgs": Array [
      Array [
        Object {
          "children": Array [],
          "parent": [Circular],
          "pos": Object {
            "end": 28,
            "start": 19,
          },
          "text": "pos arg 1",
          "type": "Text",
        },
      ],
      Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 39,
            "start": 30,
          },
          "text": "pos arg",
          "type": "CodeSpan",
        },
        Object {
          "children": Array [],
          "parent": [Circular],
          "pos": Object {
            "end": 41,
            "start": 39,
          },
          "text": " 2",
          "type": "Text",
        },
      ],
    ],
    "type": "OpCode",
  },
]
`;

exports[`macros parsing Key is defined wrongly 1`] = `
Array [
  Object {
    "keyArgs": Object {},
    "label": null,
    "opcode": "Macro",
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 15,
            "start": 0,
          },
          "text": "!Macro(@key)(@)",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 15,
        "start": 0,
      },
      "raw": "!Macro(@key)(@)",
      "type": "File",
    },
    "pos": Object {
      "end": 15,
      "start": 0,
    },
    "posArgs": Array [
      Array [
        Object {
          "children": Array [],
          "parent": [Circular],
          "pos": Object {
            "end": 11,
            "start": 7,
          },
          "text": "@key",
          "type": "Text",
        },
      ],
      Array [
        Object {
          "children": Array [],
          "parent": [Circular],
          "pos": Object {
            "end": 14,
            "start": 13,
          },
          "text": "@",
          "type": "Text",
        },
      ],
    ],
    "type": "OpCode",
  },
]
`;

exports[`table parsing Simple table 1`] = `
Array [
  Object {
    "align": Array [],
    "header": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 11,
                  "start": 1,
                },
                "text": " Column 1 ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 11,
              "start": 1,
            },
            "type": "TableCell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 22,
                  "start": 12,
                },
                "text": " Column 2 ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 22,
              "start": 12,
            },
            "type": "TableCell",
          },
        ],
        "parent": [Circular],
        "pos": Object {
          "end": 23,
          "start": 0,
        },
        "type": "TableRow",
      },
    ],
    "parent": Object {
      "children": Array [
        Object {
          "parent": [Circular],
          "pos": Object {
            "end": 99,
            "start": 0,
          },
          "text": "| Column 1 | Column 2 |
| -------- | -------- |
| Value **1** | Value \`2\` |
|Value [with](link) | |",
          "type": "Raw",
        },
      ],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 99,
        "start": 0,
      },
      "raw": "| Column 1 | Column 2 |
| -------- | -------- |
| Value **1** | Value \`2\` |
|Value [with](link) | |",
      "type": "File",
    },
    "pos": Object {
      "end": 99,
      "start": 0,
    },
    "rows": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 35,
                  "start": 25,
                },
                "text": " -------- ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 35,
              "start": 25,
            },
            "type": "TableCell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 46,
                  "start": 36,
                },
                "text": " -------- ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 46,
              "start": 36,
            },
            "type": "TableCell",
          },
        ],
        "parent": [Circular],
        "pos": Object {
          "end": 47,
          "start": 24,
        },
        "type": "TableRow",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 62,
                  "start": 49,
                },
                "text": " Value **1** ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 62,
              "start": 49,
            },
            "type": "TableCell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 70,
                  "start": 63,
                },
                "text": " Value ",
                "type": "Text",
              },
              Object {
                "parent": [Circular],
                "pos": Object {
                  "end": 73,
                  "start": 70,
                },
                "text": "2",
                "type": "CodeSpan",
              },
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 74,
                  "start": 73,
                },
                "text": " ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 74,
              "start": 63,
            },
            "type": "TableCell",
          },
        ],
        "parent": [Circular],
        "pos": Object {
          "end": 75,
          "start": 48,
        },
        "type": "TableRow",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 83,
                  "start": 77,
                },
                "text": "Value ",
                "type": "Text",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "parent": [Circular],
                    "pos": Object {
                      "end": 88,
                      "start": 84,
                    },
                    "text": "with",
                    "type": "Text",
                  },
                ],
                "href": "link",
                "parent": [Circular],
                "pos": Object {
                  "end": 95,
                  "start": 83,
                },
                "title": "",
                "type": "Link",
              },
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 96,
                  "start": 95,
                },
                "text": " ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 96,
              "start": 77,
            },
            "type": "TableCell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 98,
                  "start": 97,
                },
                "text": " ",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 98,
              "start": 97,
            },
            "type": "TableCell",
          },
        ],
        "parent": [Circular],
        "pos": Object {
          "end": 99,
          "start": 76,
        },
        "type": "TableRow",
      },
    ],
    "type": "Table",
  },
]
`;
